### ISSUE TITLE
Tasks queued behind other interrupted tasks can run in the wrong order when resolving an interruption.

### ISSUE THREAD

**Issue by borland** [OctopusDeploy/Issues#7930]
### Team

- [X] I've assigned a team label to this issue

### Severity

Low. It is unlikely a customer would experience this, and if they did they may not notice

### Version

Found in latest git master (nov 2022). Code which causes the bug dates to June 2021 or earlier.

### Latest Version

I could reproduce the problem in the latest build

### What happened?

If there is an interrupted task A (probably a deployment), with another task B behind it in the queue:

When resolving the interruption for task A, task A is supposed to resume. B should not start until A is complete.

There is a bug in the server that can cause task B to sometimes start instead; Task A does not resume until after B completes.

### Reproduction

*Pre-requisites:*

Create a project which has a manual intervention step in its process (BlockConcurrentDeployments set to true)
Configure it so a deployment is possible (create a release, etc)

*Repro:)

- Start two deployments; this will create two server tasks, referred to as A and B hereafter
- wait for task A to reach the manual intervention step

At this point, observe the task queue; you should see that task A is queued(interrupted) and task B is queued(not yet started)

- resolve the interruption on task A

What should happen:

- task A resumes and continues on with its deployment process until it reaches completion (whether it passes or fails doesn't matter here)
- Only once task A completes should task B start. Task B should reach its own interruption state, and we can resolve it as normal.

However, this can occasionally happen instead:
Teamcity build logs suggest that this happens maybe one in 10 or one in 20 times, depending on timing/luck in the server.

- rather than task A resuming, task B starts instead.
- Task B reaches its own interruption

At this point, observe the task queue; task B is queued(interrupted) and task A is queued(stuck behind task A)

### More Information

@droyad and I did some investigation, and believe we have identified the source of the problem

There are two actors (threads) here:

Thread 1: Resolving the interruption via a call to /api/interruptions/submit
Thread 2: Task queue looking for the next piece of work to do.

_Note: As at time of writing, thread 1 is the SubmitInterruptionResponder nancy responder. This is going to ported to MVC soon so it will change. The behaviour itself is not in the responder, it's just the thing that triggers it, so the port won't alter this._

Thread 1 creates a transaction, in which it
- does an ordinary write to the Interruption table; updating a row to indicate the interruption has been resolved
- writes an audit event
- does a write to the ServerTask table, updating the task.HasPendingInterruptions field to false ***
- checks for linked interruptions, and updates those (and any associated tasks) too
- commits the transaction. +++

This is interesting because there is now a time window (starting at *** and ending at +++) where the "committed" state of the server task is HasPendingInterruptions=true, but if another thread were to do a database "dirty read" with NOLOCK or similar, it would observe HasPendingInterruptions=false

Thread 2

**Comment by borland** [comment]
[7930-repro.zip](https://github.com/OctopusDeploy/Issues/files/10099166/7930-repro.zip)

**Comment by borland** [comment]
Thoughts from Rob and I:
The NOLOCK is the cause of the bug, and it does seem odd to be doing dirty reads in an area which is so important for correctness.

We couldn't see any note or reason as to why the NOLOCK was put there though. Usually things like this happen because people are attempting to fix database deadlocks, or perhaps for performance reasons; so if you are considering removing the NOLOCK, please tread carefully.
